### ISSUE TITLE
Cloudfront Distribution changes ForwardedValues Headers for Behaviours 

### ISSUE THREAD

**Issue by skorfmann** [terraform-providers/terraform-provider-aws#7867]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.11.11

### Affected Resource(s)

* aws_cloudfront_distribution

AWS Provider version: 2.1.0

### Terraform Configuration Files

```hcl
resource "aws_cloudfront_distribution" "self" {
  enabled         = true
  is_ipv6_enabled = true
  comment         = "public assets"
  price_class     = "PriceClass_100"
  aliases         = ["${aws_acm_certificate.cdn_cert.domain_name}"]
# ...
  ordered_cache_behavior {
    allowed_methods        = ["GET", "HEAD", "OPTIONS"]
    cached_methods         = ["GET", "HEAD"]
    target_origin_id       = "${local.s3_origin_id}"
    viewer_protocol_policy = "allow-all"
    compress               = true
    default_ttl            = 86400
    min_ttl                = 3600
    max_ttl                = 2592000

    path_pattern = "/foo/*"

    forwarded_values {
      query_string = false

      headers = [
        "Origin",
        "Access-Control-Request-Headers",
        "Access-Control-Request-Method",
      ]

      cookies {
        forward = "none"
      }
    }
  }
# ...
}
```

### Expected Behavior

A clean diff, since the code wasn't changed.

### Actual Behavior

There's always a diff for the `forwarwarded values headers`: 

![screenshot 2019-03-08 at 11 48 59](https://user-images.githubusercontent.com/136789/54026029-b4785780-419c-11e9-8e30-2ea40a23b8e6.png)

### Steps to Reproduce

1. `terraform plan`
1. `terraform apply`
1. `terraform plan`

**Comment by skorfmann** [comment]
Tried a downgrade of the AWS provider to `2.0.0`

apply output (apparently no actual change to the distribution, way to quick for that)

```
ordered_cache_behavior.0.forwarded_values.0.headers.0: "Access-Control-Request-Headers" => "Origin"
ordered_cache_behavior.0.forwarded_values.0.headers.1: "Access-Control-Request-Method" => "Access-Control-Request-Headers"
ordered_cache_behavior.0.forwarded_values.0.headers.2: "Origin" => "Access-Control-Request-Method"
```

state afterwards
```
"ordered_cache_behavior.0.forwarded_values.0.headers.0": "Access-Control-Request-Headers",
"ordered_cache_behavior.0.forwarded_values.0.headers.1": "Access-Control-Request-Method",
"ordered_cache_behavior.0.forwarded_values.0.headers.2": "Origin",
```

Hence, there's also a diff for each plan, but at least the apply doesn't do a full cloudfront distribution deployment now.

**Comment by jrmuir** [comment]
Some information that might be helpful.

This appears to be an issue with ordering of the item lists not matching up with what was specified, and we're seeing it with both the cookie list and header forward values.  

If we have four values applied to cookie name whitelist, e.g. `whitelisted_names = ["wp-*", "wordpress*", "MSISAuth*", "PHPSESSID"]` then every time it runs, it tries to set them to `whitelisted_names = ["MSISAuth*", "PHPSESSID", "wordpress*", "wp-*"]`.  If we adjust the source to match what it tries to set each time, it has no trouble with them because now they'll match what the provider is trying to do.  

Terraform.tfstate prior to the change shows:
```
"default_cache_behavior.0.forwarded_values.0.cookies.0.whitelisted_names.#": "4",
"default_cache_behavior.0.forwarded_values.0.cookies.0.whitelisted_names.0": "wp-*",
"default_cache_behavior.0.forwarded_values.0.cookies.0.whitelisted_names.1": "wordpress*",
"default_cache_behavior.0.forwarded_values.0.cookies.0.whitelisted_names.2": "MSISAuth*",
"default_cache_behavior.0.forwarded_values.0.cookies.0.whitelisted_names.3": "PHPSESSID",
```
After re-ordering them in our .tf file and applying, shows:
```
"default_cache_behavior.0.forwarded_values.0.cookies.0.whitelisted_names.#": "4",
"default_cache_behavior.0.forwarded_values.0.cookies.0.whitelisted_names.0": "MSISAuth*",
"default_cache_behavior.0.forwarded_values.0.cookies.0.whitelisted_names.1": "PHPSESSID",
"default_cache_behavior.0.forwarded_values.0.cookies.0.whitelisted_names.2": "wordpress*",
"default_cache_behavior.0.forwarded_values.0.cookies.0.whitelisted_names.3": "wp-*",
```
and from that point on it will run without trying to make any changes.  But if we adjust our order in our .tf file, it will continue to try to put them back to the order with MSISAuth first.

**Comment by yardensachs** [comment]
I this an AWS provider issue or a terraform issue?

**Comment by bflad** [comment]
@yardensachs this is an issue in the Terraform AWS Provider where the resource is expecting explicit ordering while the API does not. We are planning on addressing this once #8116 is merged.

**Comment by bflad** [comment]
Pull request submitted: #8150

**Comment by bflad** [comment]
The fix to ignore ordering for `headers` and cookies `whitelisted_names` has been merged and will release with version 2.5.0 of the Terraform AWS Provider, likely later today.

**Comment by nathanielks** [comment]
Thanks for your work on this @bflad!

**Comment by yardensachs** [comment]
Thanks @bflad!

**Comment by bflad** [comment]
This has been released in [version 2.5.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#250-april-05-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
